packages/@aws-cdk/aws-events/lib/event-pattern.ts

```diff
@@ -1,3 +1,315 @@
+  constructor(private readonly values: any[]) {
+    if (values.length === 0) {
+      throw new Error('anythingBut matchers must be non-empty lists');
+    }
+
+    if (!(this.allNumbers() || this.allStrings())) {
+      throw new Error(`anythingBut matchers must be lists that contain only strings or only numbers. Got: ${values}`);
+    }
+  }
+
+  toEventBridgeMatcher() {
+    return [{ 'anything-but': this.values }];
+  }
+
+  private allStrings(): boolean {
+    return this.values.every(v => typeof (v) === 'string');
+  }
+
+  private allNumbers(): boolean {
+    return this.values.every(v => typeof (v) === 'number');
+  }
+}
+
+class AnythingButPrefixMatcher implements IMatcher {
+  constructor(private readonly prefix: string) {}
+
+  toEventBridgeMatcher(): any[] {
+    return [{ 'anything-but': { prefix: this.prefix } }];
+  }
+}
+
+class LessThanMatcher implements INumericMatcher {
+  constructor(private readonly value: number) { }
+
+  toEventBridgeMatcher() {
+    return [{ numeric: ['<', this.value] }];
+  }
+}
+
+class LessThanOrEqualMatcher implements INumericMatcher {
+  constructor(private readonly value: number) { }
+
+  toEventBridgeMatcher() {
+    return [{ numeric: ['<=', this.value] }];
+  }
+}
+
+class GreaterThanMatcher implements INumericMatcher {
+  constructor(private readonly value: number) { }
+
+  toEventBridgeMatcher() {
+    return [{ numeric: ['>', this.value] }];
+  }
+}
+
+class GreaterThanOrEqualMatcher implements INumericMatcher {
+  constructor(private readonly value: number) { }
+
+  toEventBridgeMatcher() {
+    return [{ numeric: ['>=', this.value] }];
+  }
+}
+
+class EqualMatcher implements INumericMatcher {
+  constructor(private readonly value: number) { }
+
+  toEventBridgeMatcher() {
+    return [{ numeric: ['=', this.value] }];
+  }
+}
+
+class IntervalMatcher implements IMatcher {
+  constructor(
+    private readonly lower: number,
+    private readonly uppper: number,
+  ) {
+    if (lower > uppper) {
+      throw new Error(`Invalid interval: [${lower}, ${uppper}]`);
+    }
+  }
+
+  toEventBridgeMatcher() {
+    return [{ numeric: ['>=', this.lower, '<=', this.uppper] }];
+  }
+}
+
+class NumericMatcher implements IMatcher {
+  constructor(private readonly matchers: INumericMatcher[]) {
+    if (matchers.length === 0) {
+      throw new Error('numeric matchers must be non-empty lists');
+    }
+  }
+
+  toEventBridgeMatcher(): any {
+    return [{
+      numeric: this.matchers
+        .flatMap(m => m.toEventBridgeMatcher())
+        .flatMap(m => m.numeric),
+    }];
+  }
+}
+
+class AnyOfMatcher implements IMatcher {
+  constructor(private readonly matchers: IMatcher[]) {
+    if (matchers.length === 0) {
+      throw new Error('anyOf matchers must be non-empty lists');
+    }
+  }
+
+  toEventBridgeMatcher(): any[] {
+    return this.matchers.flatMap(m => m.toEventBridgeMatcher());
+  }
+}
+
+/**
+ * A collection of matchers for constructing event patterns
+ */
+export class Matchers {
```

Comment:
I think I'd prefer `Match` as a class name? It'll look nicer in code? As in, `Match.anythingBut(...)` ?